### PR TITLE
CUDA Nonbonded Precomputed

### PR DESCRIPTION
**Pull request by proteneer**
This is a pretty simple PR, and mirrors the nonbonded jax implementation identically.

Note: This class should be used for small-molecule intramolecular interactions only (ligands in particular). This class should not be used for the cancellation of nans since the interpolation of `q_ij, sig_ij, eps_ij` is done on python side and is *not* bit-wise cancellable with the exclusions.

### CODE REVIEW

File: timemachine/cpp/src/nonbonded_precomputed.cu [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    unsigned long long *d_du_dx,
+    unsigned long long *d_du_dp,
+    unsigned long long *d_du_dl,
+    unsigned long long *d_u,
+    cudaStream_t stream) {
+
+    if (P != 3 * B_) {
+        throw std::runtime_error(
+            "NonbondedPairListPrecomputed::execute_device(): expected P == 2*B, got P=" + std::to_string(P) +
+            ", 2*B=" + std::to_string(2 * B_));
```
Comment (badisa) [inline timemachine/cpp/src/nonbonded_precomputed.cu:61]:
```suggestion
            "NonbondedPairListPrecomputed::execute_device(): expected P == 3*B, got P=" + std::to_string(P) +
            ", 3*B=" + std::to_string(3 * B_));
```

File: timemachine/potentials/nonbonded.py [hunk 2]
```diff
@@ -326,6 +334,7 @@ def nonbonded_v3_on_precomputed_pairs(conf, params, box, pairs, offsets, beta: f
     conf: N,3
     params: P,3 (q_ij, s_ij, e_ij)
     pairs: P,2 (i,j)
+    offsets: P, 2 (w,) distance offsets for electrostatic and lj interactions
```
Comment (proteneer) [inline timemachine/potentials/nonbonded.py:337]: note to self: fix

**Review by badisa — APPROVED** [review]

File: tests/nonbonded/test_nonbonded_precomputed.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    ixn_group_size,
+    precision,
+    rtol,
+    atol,
+    cutoff,
+    beta,
+    example_nonbonded_potential,
+    example_conf,
+    example_box,
+    rng: np.random.Generator,
```
Comment (mcwitt) [inline tests/nonbonded/test_nonbonded_precomputed.py:50]: nit: unused argument

**Review by mcwitt — APPROVED** [review]

File: tests/nonbonded/test_nonbonded_precomputed.py (hunk 3 above)
Comment (proteneer) [reply tests/nonbonded/test_nonbonded_precomputed.py:50]: the conftest setup doesn't let me rename these - and I don't think I can just remove them.

Comment (proteneer) [reply tests/nonbonded/test_nonbonded_precomputed.py:49]: nevermind - I was just confused about how the conftest works. Thanks for resolving this offline.

**Pull request merged**